programs/clearing_house/src/math/spot_withdraw.rs

```diff
@@ -0,0 +1,176 @@
+use solana_program::msg;
+
+use crate::error::{ClearingHouseResult, ErrorCode};
+use crate::math::casting::{cast, cast_to_u128};
+use crate::math::constants::QUOTE_SPOT_MARKET_INDEX;
+use crate::math::safe_math::SafeMath;
+
+use crate::math::spot_balance::get_token_amount;
+use crate::state::spot_market::{SpotBalanceType, SpotMarket};
+use crate::state::user::SpotPosition;
+use crate::validate;
+
+pub fn calculate_min_deposit_token(
+    deposit_token_twap: u128,
+    withdraw_guard_threshold: u128,
+) -> ClearingHouseResult<u128> {
+    let min_deposit_token = deposit_token_twap
+        .safe_sub((deposit_token_twap / 5).max(withdraw_guard_threshold.min(deposit_token_twap)))?;
+
+    Ok(min_deposit_token)
+}
+
+pub fn calculate_max_borrow_token(
+    deposit_token_amount: u128,
+    borrow_token_twap: u128,
+    withdraw_guard_threshold: u128,
+) -> ClearingHouseResult<u128> {
+    let max_borrow_token = withdraw_guard_threshold.max(
+        (deposit_token_amount / 6)
+            .max(borrow_token_twap.safe_add(borrow_token_twap / 5)?)
+            .min(deposit_token_amount.safe_sub(deposit_token_amount / 5)?),
+    ); // between ~15-80% utilization with friction on twap
+
+    Ok(max_borrow_token)
+}
+
+pub fn check_user_exception_to_withdraw_limits(
+    spot_market: &SpotMarket,
+    spot_position: Option<&SpotPosition>,
+    token_amount_withdrawn: Option<u128>,
+) -> ClearingHouseResult<bool> {
+    // allow a smaller user in QUOTE_SPOT_MARKET_INDEX to bypass and withdraw their principal
+    let mut valid_user_withdraw = false;
+    if let Some(spot_position) = spot_position {
+        if spot_position.market_index == QUOTE_SPOT_MARKET_INDEX
+            && spot_position.cumulative_deposits >= 0
+        {
+            validate!(
```

Comment:
should we just return false here instead of throw an error?

Reply:
yup changing